File: debuglog-annotations/src/main/java/it/sephiroth/android/library/debuglog/ParamsLogger.java
package it.sephiroth.android.library.debuglog;

import java.lang.reflect.Array;
import java.util.Arrays;

@SuppressWarnings("unused")
public class ParamsLogger {
    private int paramIndex = 0;
    private final StringBuilder paramsList = new StringBuilder();
    private final String methodName;
    private final String divider = ", ";
    private final String tag;
    private final int debugArguments;
    private final boolean printArguments;

    public ParamsLogger(String tag, String methodName, int debugType) {
        this.tag = tag;
        this.methodName = methodName;
        this.debugArguments = debugType;
        this.printArguments = true; //debugArguments != DebugArguments.NONE;
    }

    public ParamsLogger(String tag, String methodName) {
        this(tag, methodName, DebugArguments.FULL);
    }

    public ParamsLogger append(String name) {
        if (!printArguments) return this;
        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(name);
        return this;
    }

    public ParamsLogger append(String name, int val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, boolean val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, short val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, byte val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, char val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, long val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, double val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, float val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
        return this;
    }

    public ParamsLogger append(String name, Object val) {
        if (!printArguments) return this;

        if (paramIndex++ != 0) {
            paramsList.append(divider);
        }

        if (null != val) {
            if (debugArguments == DebugArguments.FULL) {
                if (val.getClass().isArray()) {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, arrayToString(val)));
                } else {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
                }
            } else if (debugArguments == DebugArguments.SHORT) {
                if (val.getClass().isArray()) {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, arrayToHashCode(val)));
                } else if (val instanceof String || val instanceof Enum) {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, val));
                } else {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, objectToHashCode(val)));
                }
            }
        } else {
            paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, "null"));
        }

        return this;
    }

    static String objectToHashCode(Object val) {
        if (val == null) return "null";
        return val.getClass().getName() + "#" + System.identityHashCode(val);
    }

    static String arrayToHashCode(Object val) {
        if (val == null) return "null";
        return val.getClass().getCanonicalName() + "[size=" + Array.getLength(val) + "]";
    }

    private String arrayToString(Object val) {
        if (!(val instanceof Object[])) {
            if (val instanceof int[]) {
                return Arrays.toString((int[]) val);
            } else if (val instanceof char[]) {
                return Arrays.toString((char[]) val);
            } else if (val instanceof boolean[]) {
                return Arrays.toString((boolean[]) val);
            } else if (val instanceof byte[]) {
                return Arrays.toString((byte[]) val);
            } else if (val instanceof long[]) {
                return Arrays.toString((long[]) val);
            } else if (val instanceof double[]) {
                return Arrays.toString((double[]) val);
            } else if (val instanceof float[]) {
                return Arrays.toString((float[]) val);
            } else if (val instanceof short[]) {
                return Arrays.toString((short[]) val);
            } else {
                return "Unknown type array";
            }
        } else {
            return Arrays.deepToString((Object[]) val);
        }
    }

    public void print(int level) {
        DebugLogger.DEFAULT_IMPL.logEnter(level, tag, methodName, paramsList.toString());
    }

    static final String PARAMETER_PRINT_FORMAT = "%s=\"%s\"";
}

File: asm-debuglog-runtime/src/main/java/it/sephiroth/android/library/asm/runtime/debuglog/ParamsLogger.kt
package it.sephiroth.android.library.asm.runtime.debuglog

import it.sephiroth.android.library.asm.runtime.debuglog.DebugLogger.instance
import it.sephiroth.android.library.asm.runtime.debuglog.annotations.DebugArguments

@Suppress("unused")
class ParamsLogger(
    private val tag: String,
    private val methodName: String,
    private val debugArguments: Int
) {
    private var paramIndex = 0
    private val paramsList = StringBuilder()
    private val divider = ", "
    private val printArguments = true

    fun append(name: String?): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(name)
        return this
    }

    fun append(name: String?, value: Int): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Boolean): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Short): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Byte): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Char): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Long): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Double): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Float): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
        return this
    }

    fun append(name: String?, value: Any?): ParamsLogger {
        if (!printArguments) return this
        if (paramIndex++ != 0) {
            paramsList.append(divider)
        }
        if (null != value) {
            if (debugArguments == DebugArguments.FULL) {
                if (value.javaClass.isArray) {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, arrayToString(value)))
                } else {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
                }
            } else if (debugArguments == DebugArguments.SHORT) {
                if (value.javaClass.isArray) {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, arrayToHashCode(value)))
                } else if (value is String || value is Enum<*>) {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, value))
                } else {
                    paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, objectToHashCode(value)))
                }
            }
        } else {
            paramsList.append(String.format(PARAMETER_PRINT_FORMAT, name, "null"))
        }
        return this
    }

    fun print(level: Int) {
        instance.logEnter(level, tag, methodName, paramsList.toString())
    }

    companion object {
        @JvmStatic
        fun objectToHashCode(value: Any?): String {
            return if (value == null) "null" else value::class.java.name + "#" + System.identityHashCode(value)
        }

        @JvmStatic
        fun arrayToHashCode(value: Any?): String {
            return if (value == null) "null" else value::class.java.name + "[size=" + java.lang.reflect.Array.getLength(value) + "]"
        }

        @JvmStatic
        internal fun arrayToString(value: Any): String {
            return when (value) {
                is Array<*> -> value.contentDeepToString()
                is ByteArray -> value.contentToString()
                is ShortArray -> value.contentToString()
                is IntArray -> value.contentToString()
                is LongArray -> value.contentToString()
                is FloatArray -> value.contentToString()
                is DoubleArray -> value.contentToString()
                is CharArray -> value.contentToString()
                is BooleanArray -> value.contentToString()

                // Experimental types
//            is UByteArray -> value.contentToString()
//            is UShortArray -> value.contentToString()
//            is UIntArray -> value.contentToString()
//            is ULongArray -> value.contentToString()
                else -> value.toString()
            }
        }

        const val PARAMETER_PRINT_FORMAT = "%s=\"%s\""
    }
}
